integration/gateway/gateway_discovery_test.go

```diff
@@ -0,0 +1,365 @@
+		Expect(err).NotTo(HaveOccurred())
+
+		preparedTransaction := endorseResponse.GetPreparedTransaction()
+		preparedTransaction.Signature, err = signer.Sign(preparedTransaction.Payload)
+		Expect(err).NotTo(HaveOccurred())
+
+		verifyEndorsers(endorseResponse, expectedEndorsers)
+
+		submitRequest := &gateway.SubmitRequest{
+			TransactionId:       transactionID,
+			ChannelId:           channel,
+			PreparedTransaction: preparedTransaction,
+		}
+		_, err = gatewayClient.Submit(ctx, submitRequest)
+		Expect(err).NotTo(HaveOccurred())
+
+		idBytes, err := signer.Serialize()
+		Expect(err).NotTo(HaveOccurred())
+
+		statusRequest := &gateway.CommitStatusRequest{
+			ChannelId:     "testchannel",
+			Identity:      idBytes,
+			TransactionId: transactionID,
+		}
+		statusRequestBytes, err := proto.Marshal(statusRequest)
+		Expect(err).NotTo(HaveOccurred())
+
+		signature, err := signer.Sign(statusRequestBytes)
+		Expect(err).NotTo(HaveOccurred())
+
+		signedStatusRequest := &gateway.SignedCommitStatusRequest{
+			Request:   statusRequestBytes,
+			Signature: signature,
+		}
+
+		_, err = gatewayClient.CommitStatus(ctx, signedStatusRequest)
+		Expect(err).NotTo(HaveOccurred())
+	}
+
+	evaluateTransaction := func(
+		gatewayClient gateway.GatewayClient,
+		signer *nwo.SigningIdentity,
+		channel string,
+		chaincode string,
+		transactionName string,
+		arguments []string,
+		transientData map[string][]byte,
+	) *peer.Response {
+		args := [][]byte{}
+		for _, arg := range arguments {
+			args = append(args, []byte(arg))
+		}
+		proposedTransaction, transactionID := NewProposedTransaction(
+			signer,
+			channel,
+			chaincode,
+			transactionName,
+			transientData,
+			args...,
+		)
+
+		evaluateRequest := &gateway.EvaluateRequest{
+			TransactionId:       transactionID,
+			ChannelId:           channel,
+			ProposedTransaction: proposedTransaction,
+		}
+
+		ctx, cancel := context.WithTimeout(context.Background(), network.EventuallyTimeout)
+		defer cancel()
+
+		evaluateResponse, err := gatewayClient.Evaluate(ctx, evaluateRequest)
+		Expect(err).NotTo(HaveOccurred())
+
+		return evaluateResponse.GetResult()
+	}
+
+	It("SBE policy should cause extra endorsers to be selected", func() {
+		conn := network.PeerClientConn(org1Peer0)
+		defer conn.Close()
+		gatewayClient := gateway.NewGatewayClient(conn)
+		signingIdentity := network.PeerUserSigner(org1Peer0, "User1")
+
+		// add org2 to SBE policy - requires endorsement from org3 peer (chaincode policy)
+		submitTransaction(
+			gatewayClient,
+			signingIdentity,
+			"testchannel",
+			"sbecc",
+			"addorgs",
+			[]string{"pub", "Org2MSP"},
+			nil,
+			[]string{org3Peer0.ID()},
+		)
+
+		// add org1 to SBE policy - requires endorsement from org2 & org3 peers
+		submitTransaction(
+			gatewayClient,
+			signingIdentity,
+			"testchannel",
+			"sbecc",
+			"addorgs",
+			[]string{"pub", "Org1MSP"},
+			nil,
+			[]string{org2Peer0.ID(), org3Peer0.ID()},
+		)
+
```

Comment:
Why would this still need an endorsement from `org3` (i.e., chaincode level endorsement?) Am I missing something or in the production code, it causes unnecessary addition of chaincode policy?

Reply:
Because when this function was added to the gateway, there was no way of expressing the fact that the chaincode policy should be ignored in the ChaincodeInterest structure.  However, support for this has subsequently been added to the discovery service (https://github.com/hyperledger/fabric/commit/e5e623d92b74772f30246ecd0ebfd0274ae4bae2) and so we can add this into the gateway.

I'll raise an issue to track this work.  
Thanks.

Reply:
Opened https://github.com/hyperledger/fabric/issues/2856